util/Pending.h

```diff
@@ -38,62 +38,84 @@ namespace Pending {
+                if (pending.pending->valid()) {
+                    DebugLogger() << "Dont care for result of parsing \"" << pending.filename << "\". Have to get() once to release shared state in pending future.";
+                    pending.pending->get(); // needs to be called once to release state
                 }
-                DebugLogger() << "Dont care for result of parsing \"" << pending->filename << "\". Was already released.";
-                pending = boost::none;
+                DebugLogger() << "Dont care for result of parsing \"" << pending.filename << "\". Was already released.";
                 return boost::none;
             }
-            DebugLogger() << "Retrieve result of parsing \"" << pending->filename << "\".";
-            auto x = std::move(pending->pending->get());
-            pending = boost::none;
+            DebugLogger() << "Retrieve result of parsing \"" << pending.filename << "\".";
+            auto x = std::move(pending.pending->get());
+            DebugLogger() << "Retrieved result of parsing \"" << pending.filename << "\".";
             return std::move(x);
         } catch (const std::exception& e) {
-            ErrorLogger() << "Parsing of \"" << pending->filename << "\" failed with error: " << e.what();
-            pending = boost::none;
+            ErrorLogger() << "Parsing of \"" << pending.filename << "\" failed with error: " << e.what();
         }
 
         return boost::none;
     }
+    /** Wait for the \p pending parse to complete.  Set pending to boost::none
+        and return the parsed T. Destroys the shared state in the wrapped std::future.
+        Return boost::none on errors.*/
+    template <typename T>
+    boost::optional<T> WaitForPending(boost::optional<Pending<T>>& pending, bool do_not_care_about_result = false) {
+        if (!pending)
+            return boost::none;
+        std::lock_guard<std::mutex> lock(pending->m_mutex);
+        if (!pending || !(pending->pending)) {
+            return boost::none;
+        }
+        if (auto tt = WaitForPendingUnlocked(std::move(*pending), do_not_care_about_result)) {
+            pending = boost::none; // XXX
+            return tt;
+        } else {
+            pending = boost::none;
+            return boost::none;
+        }
+    }
 
     /** If there is a pending parse, wait for it and swap it with the stored
         value.  Return the stored value.*/
     template <typename T>
     T& SwapPending(boost::optional<Pending<T>>& pending, T& stored) {
         if (pending) {
-            std::lock_guard<std::recursive_mutex> lock(pending->m_mutex);
-            if (auto tt = WaitForPending(pending))
+            std::lock_guard<std::mutex> lock(pending->m_mutex);
+            if (!pending) {
+                return stored;
+            }
+            if (auto tt = WaitForPendingUnlocked(std::move(*pending))) {
                 std::swap(*tt, stored);
+                // FIXME always setting to none is probably wrong
```

Comment:
Is this FIXME important?

Reply:
Before I did not have time for doing another thorough code analysis of the code paths of the old code (at b2f465fe65774a7012c234c53a4b3b84066b3a08) and i wanted to see first that it works for vezzra.

So, it looks for all non-exceptional cases that pending was always none for both SwapPending and WaitForPending, so this should be ok.

There is some throw for std::future_status::deferred state, but AFAIU that state should not happen.